File: agent/db/sql/integrity.py
"""
agent/db/sql/integrity.py — Этап 5E-S S7: tamper-evident integrity
journal (mandate §24-§27).

DESIGNED + unit-tested this pass, NOT wired into any production write
path (see SECURITY_ARCHITECTURE.md §21) — the primitives (canonical
serialization, hash chain, checkpoint comparison) are proven correct
in isolation first; integrating them into every canonical INSERT is
the next continuation.

This is a SECOND, independent wall behind SQL privileges (security_
grants.py) and immutability triggers (security_triggers.py) — it
exists specifically for the case those two cannot fully cover: an
attacker with admin-level bypass (SUPER/root) who disables a trigger
or a database restored from an old-but-internally-valid backup
(mandate §22/§27 — "rollback attack"). The chain DETECTS such tampering
after the fact; it does not prevent or restore (SECURITY_ARCHITECTURE.md
§14 says this plainly).

Design choices, per the mandate's own cautions against over-
engineering (§26): ONE chain per entity_type (not one global chain —
avoids serializing every canonical write through a single lock; "YANDI
сейчас не high-frequency trading system"), no Merkle-tree framework,
HMAC-SHA256 (not a hand-rolled construction).
"""
from __future__ import annotations

import hashlib
import hmac
import json
import os
import time
from datetime import datetime
from typing import Any, Dict, List, Optional, Tuple, Union

INTEGRITY_FORMAT_VERSION = 1
GENESIS_HASH = "0" * 64


class IntegrityError(Exception):
    """Base for this module's errors."""


class UnsupportedFieldType(IntegrityError):
    """A field value passed to canonicalize_record() isn't one of the
    explicitly-supported JSON-safe types — raised rather than silently
    guessing a serialization (mandate §25: no ambiguous formatting)."""


def canonicalize_value(value: Any) -> Any:
    """Normalizes ONE field value to a deterministic JSON-safe form.

    - None -> null (explicit, mandate §25).
    - bool/int/str -> unchanged (json's own serialization of these is
      already deterministic and platform-independent).
    - float -> a FIXED-PRECISION STRING ("%.6f"), never a raw float —
      avoids any ambiguity between "the same number, formatted
      differently" being treated as a hash mismatch, and sidesteps
      relying on Python's float repr staying stable forever.
    - datetime -> REJECTED. Caller must convert to an integer (e.g.
      microseconds since epoch) before calling — mandate §25: "integer
      time precision / normalized datetime", not a raw datetime object
      whose serialization the caller didn't explicitly choose.
    """
    if value is None or isinstance(value, bool) or isinstance(value, int) or isinstance(value, str):
        return value
    if isinstance(value, float):
        return f"{value:.6f}"
    if isinstance(value, datetime):
        raise UnsupportedFieldType(
            "datetime objects are not canonical-serializable directly — convert to "
            "an integer (e.g. microseconds since epoch) before calling canonicalize_record()"
        )
    raise UnsupportedFieldType(f"unsupported type for canonical serialization: {type(value)!r}")


def canonicalize_record(
    fields: Dict[str, Any], *, entity_type: str, entity_id: Union[str, int],
    format_version: int = INTEGRITY_FORMAT_VERSION,
) -> bytes:
    """Deterministic, versioned byte representation of one canonical
    record (mandate §25). Same logical record -> byte-identical output
    on every call, on every platform: UTF-8, sort_keys, explicit
    separators (no incidental whitespace differences), an explicit
    format_version field so a future serializer change doesn't break
    the ability to verify OLD history (mandate: "Изменение serializer
    в будущем не должно уничтожать возможность проверить старую
    историю" — old events keep re-verifying against THEIR OWN recorded
    format_version, not whatever version canonicalize_record() defaults
    to today)."""
    envelope = {
        "format_version": format_version,
        "entity_type": entity_type,
        "entity_id": str(entity_id),
        "fields": {k: canonicalize_value(v) for k, v in fields.items()},
    }
    return json.dumps(envelope, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode("utf-8")


def payload_hash(canonical_bytes: bytes) -> str:
    return hashlib.sha256(canonical_bytes).hexdigest()


def compute_event_hash(
    integrity_key: bytes, format_version: int, entity_type: str, entity_id: Union[str, int],
    payload_hash_hex: str, previous_event_hash_hex: str,
) -> str:
    """mandate §26: event_hash = HMAC-SHA256(K_integrity, format_version
    || entity_type || entity_id || payload_hash || previous_event_hash).
    K_integrity is a DEDICATED key (keys.py::derive_integrity_key()),
    never an encryption DEK/KEK."""
    if not integrity_key:
        raise IntegrityError("compute_event_hash() called with no integrity key.")
    message = f"{format_version}|{entity_type}|{entity_id}|{payload_hash_hex}|{previous_event_hash_hex}".encode("utf-8")
    return hmac.new(integrity_key, message, hashlib.sha256).hexdigest()


def append_event(
    integrity_key: bytes, chain: List[Dict[str, Any]], *, entity_type: str,
    entity_id: Union[str, int], fields: Dict[str, Any], format_version: int = INTEGRITY_FORMAT_VERSION,
) -> Dict[str, Any]:
    """Builds the NEXT event for one entity_type's chain — does not
    mutate `chain` (caller decides how/where to persist the returned
    event, e.g. a future integrity_event table)."""
    canonical = canonicalize_record(fields, entity_type=entity_type, entity_id=entity_id, format_version=format_version)
    p_hash = payload_hash(canonical)
    previous_event_hash = chain[-1]["event_hash"] if chain else GENESIS_HASH
    seq = (chain[-1]["seq"] + 1) if chain else 1
    event_hash = compute_event_hash(integrity_key, format_version, entity_type, entity_id, p_hash, previous_event_hash)
    return {
        "seq": seq,
        "entity_type": entity_type,
        "entity_id": str(entity_id),
        "format_version": format_version,
        "payload_hash": p_hash,
        "previous_event_hash": previous_event_hash,
        "event_hash": event_hash,
    }


def verify_chain(integrity_key: bytes, events: List[Dict[str, Any]]) -> Tuple[bool, List[str]]:
    """Walks a sequence of events and confirms: (1) each event's
    previous_event_hash actually equals the PRECEDING event's real
    event_hash (catches a deleted middle event — the gap breaks this
    link; catches reordering — swapping two events breaks it too), and
    (2) each event's own event_hash correctly recomputes from its
    stored payload_hash/metadata (catches a tampered event_hash or
    metadata field that wasn't recomputed consistently).

    Does NOT by itself catch "attacker modified the row's data AND
    recomputed a consistent payload_hash/event_hash" — that requires
    the integrity_key, which an attacker without it cannot do; but if
    you want to verify actual CURRENT row content against a stored
    event, use verify_record_against_event() below."""
    problems: List[str] = []
    previous_event_hash = GENESIS_HASH
    for event in events:
        if event["previous_event_hash"] != previous_event_hash:
            problems.append(
                f"seq={event.get('seq')}: previous_event_hash mismatch "
                f"(expected {previous_event_hash}, got {event['previous_event_hash']}) "
                f"— a prior event was deleted, reordered, or replaced"
            )
        expected_hash = compute_event_hash(
            integrity_key, event["format_version"], event["entity_type"], event["entity_id"],
            event["payload_hash"], event["previous_event_hash"],
        )
        if event["event_hash"] != expected_hash:
            problems.append(f"seq={event.get('seq')}: event_hash does not match recomputation — tampered")
        previous_event_hash = event["event_hash"]
    return (len(problems) == 0, problems)


def verify_record_against_event(
    fields: Dict[str, Any], event: Dict[str, Any], *, entity_type: str, entity_id: Union[str, int],
) -> bool:
    """Recomputes the canonical payload_hash from ACTUAL CURRENT field
    values and compares against what the stored event claims — this is
    the check that actually catches "the row's content was modified in
    place" (T12/T18/T21), independent of whether the chain itself still
    looks internally consistent."""
    canonical = canonicalize_record(
        fields, entity_type=entity_type, entity_id=entity_id, format_version=event["format_version"],
    )
    return payload_hash(canonical) == event["payload_hash"]


# ============================================================
# External checkpoint (mandate §27 — rollback detection).
# ============================================================

def make_checkpoint(entity_type: str, events: List[Dict[str, Any]], path: str) -> Dict[str, Any]:
    """Writes {entity_type, seq, event_hash, created_at} to `path`
    (mandate: outside SQL, a protected local filesystem path) —
    atomically (write to a temp file, fsync, os.replace) so a crash
    mid-write never leaves a half-written checkpoint. Contains no
    knowledge (mandate §27: "Checkpoint: не содержит knowledge")."""
    if not events:
        raise ValueError("cannot checkpoint an empty chain")
    latest = events[-1]
    checkpoint = {
        "entity_type": entity_type,
        "seq": latest["seq"],
        "event_hash": latest["event_hash"],
        "created_at": int(time.time()),
    }
    parent = os.path.dirname(path) or "."
    os.makedirs(parent, mode=0o700, exist_ok=True)
    tmp_path = f"{path}.tmp"
    with open(tmp_path, "w", encoding="utf-8") as f:
        json.dump(checkpoint, f, sort_keys=True)
        f.flush()
        os.fsync(f.fileno())
    os.replace(tmp_path, path)
    return checkpoint


def load_checkpoint(path: str) -> Optional[Dict[str, Any]]:
    if not os.path.exists(path):
        return None
    with open(path, encoding="utf-8") as f:
        return json.load(f)


def check_for_rollback(
    entity_type: str, current_events: List[Dict[str, Any]], checkpoint_path: str,
) -> Tuple[str, Optional[str]]:
    """Returns (status, detail). status is one of:
        'no_checkpoint'      — nothing to compare against yet (first run)
        'ok'                 — DB matches the last confirmed checkpoint
        'ahead'              — DB has progressed since the checkpoint (normal)
        'ROLLBACK_SUSPECTED' — DB is BEHIND a previously-confirmed
                               checkpoint, or diverged at the same
                               sequence number — mandate §27/§28: MUST
                               NOT proceed with canonical writes
                               silently when this is returned."""
    checkpoint = load_checkpoint(checkpoint_path)
    if checkpoint is None:
        return ("no_checkpoint", None)

    if not current_events:
        return ("ROLLBACK_SUSPECTED", "DB has zero events for an entity_type with a prior checkpoint")

    latest = current_events[-1]
    if latest["seq"] < checkpoint["seq"]:
        return ("ROLLBACK_SUSPECTED", f"DB seq {latest['seq']} < checkpoint seq {checkpoint['seq']}")
    if latest["seq"] == checkpoint["seq"] and latest["event_hash"] != checkpoint["event_hash"]:
        return ("ROLLBACK_SUSPECTED", "same sequence number but different hash — history has diverged")
    if latest["seq"] == checkpoint["seq"]:
        return ("ok", None)
    return ("ahead", None)

File: rustlib/yandi_db/src/integrity.rs
//! Журнал целостности с цепочкой хешей — перенос `agent/db/sql/integrity.py`: детерминированная сериализация записи, HMAC-SHA256 цепочка, контрольная точка вне базы,
//! обнаружение отката. Формат байт-в-байт как у Python: события и контрольные точки, созданные одной реализацией, проверяет другая.
use std::fs;
use std::io::Write;
use std::path::Path;

use hmac::{Hmac, Mac};
use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};

pub const INTEGRITY_FORMAT_VERSION: i64 = 1;
pub const GENESIS_HASH: &str = "0000000000000000000000000000000000000000000000000000000000000000";

pub type R<T> = Result<T, String>;

/// Значение поля в каноническом виде: None/bool/int/str — как есть, float — строка `%.6f`; остальное отвергается.
pub fn canonicalize_value(v: &Value) -> R<Value> {
    match v {
        Value::Null | Value::Bool(_) | Value::String(_) => Ok(v.clone()),
        Value::Number(n) if n.is_i64() || n.is_u64() => Ok(v.clone()),
        Value::Number(n) => Ok(Value::String(format!("{:.6}", n.as_f64().unwrap_or(0.0)))),
        other => Err(format!("UnsupportedFieldType: unsupported type for canonical serialization: {}", match other {
            Value::Array(_) => "<class 'list'>",
            _ => "<class 'dict'>",
        })),
    }
}

/// `json.dumps(..., ensure_ascii=False)` строки.
fn dumps_str(s: &str, out: &mut String) {
    out.push('"');
    for ch in s.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            '\u{8}' => out.push_str("\\b"),
            '\u{c}' => out.push_str("\\f"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
}

fn dumps_scalar(v: &Value, out: &mut String) {
    match v {
        Value::Null => out.push_str("null"),
        Value::Bool(b) => out.push_str(if *b { "true" } else { "false" }),
        Value::Number(n) => out.push_str(&n.to_string()),
        Value::String(s) => dumps_str(s, out),
        _ => unreachable!("канонизация пропускает только скаляры"),
    }
}

/// Детерминированные байты записи: ключи отсортированы, разделители `,` и `:`, UTF-8 без экранирования не-ASCII.
pub fn canonicalize_record(fields: &Map<String, Value>, entity_type: &str, entity_id: &str, format_version: i64) -> R<Vec<u8>> {
    let mut sorted: Vec<(&String, Value)> = Vec::new();
    for (k, v) in fields {
        sorted.push((k, canonicalize_value(v)?));
    }
    sorted.sort_by(|a, b| a.0.cmp(b.0));
    let mut s = String::from("{");
    // корневые ключи по алфавиту: entity_id, entity_type, fields, format_version
    s.push_str("\"entity_id\":");
    dumps_str(entity_id, &mut s);
    s.push_str(",\"entity_type\":");
    dumps_str(entity_type, &mut s);
    s.push_str(",\"fields\":{");
    for (i, (k, v)) in sorted.iter().enumerate() {
        if i > 0 {
            s.push(',');
        }
        dumps_str(k, &mut s);
        s.push(':');
        dumps_scalar(v, &mut s);
    }
    s.push_str(&format!("}},\"format_version\":{format_version}}}"));
    Ok(s.into_bytes())
}

pub fn payload_hash(canonical: &[u8]) -> String {
    Sha256::digest(canonical).iter().map(|b| format!("{b:02x}")).collect()
}

pub fn compute_event_hash(key: &[u8], format_version: i64, entity_type: &str, entity_id: &str, payload_hash_hex: &str, previous_hex: &str) -> R<String> {
    if key.is_empty() {
        return Err("IntegrityError: compute_event_hash() called with no integrity key.".into());
    }
    let mut m = <Hmac<Sha256> as Mac>::new_from_slice(key).map_err(|e| e.to_string())?;
    m.update(format!("{format_version}|{entity_type}|{entity_id}|{payload_hash_hex}|{previous_hex}").as_bytes());
    Ok(m.finalize().into_bytes().iter().map(|b| format!("{b:02x}")).collect())
}

fn s<'a>(e: &'a Value, k: &str) -> &'a str {
    e.get(k).and_then(|v| v.as_str()).unwrap_or("")
}

/// Следующее событие цепочки одного `entity_type` (цепочку не изменяет).
pub fn append_event(key: &[u8], chain: &[Value], entity_type: &str, entity_id: &str, fields: &Map<String, Value>, format_version: i64) -> R<Value> {
    let canonical = canonicalize_record(fields, entity_type, entity_id, format_version)?;
    let p = payload_hash(&canonical);
    let prev = chain.last().map(|e| s(e, "event_hash").to_string()).unwrap_or_else(|| GENESIS_HASH.to_string());
    let seq = chain.last().and_then(|e| e.get("seq").and_then(|v| v.as_i64())).map(|x| x + 1).unwrap_or(1);
    let h = compute_event_hash(key, format_version, entity_type, entity_id, &p, &prev)?;
    Ok(json!({"seq": seq, "entity_type": entity_type, "entity_id": entity_id, "format_version": format_version, "payload_hash": p, "previous_event_hash": prev, "event_hash": h}))
}

/// Проверка цепочки: `(целая ли, найденные проблемы)`.
pub fn verify_chain(key: &[u8], events: &[Value]) -> R<(bool, Vec<String>)> {
    let mut problems = Vec::new();
    let mut prev = GENESIS_HASH.to_string();
    for e in events {
        if s(e, "previous_event_hash") != prev {
            problems.push(format!(
                "seq={}: previous_event_hash mismatch (expected {}, got {}) — a prior event was deleted, reordered, or replaced",
                e.get("seq").map(py_str).unwrap_or_else(|| "None".into()),
                prev,
                s(e, "previous_event_hash")
            ));
        }
        let expected = compute_event_hash(key, e.get("format_version").and_then(|v| v.as_i64()).unwrap_or(0), s(e, "entity_type"), s(e, "entity_id"), s(e, "payload_hash"), s(e, "previous_event_hash"))?;
        if s(e, "event_hash") != expected {
            problems.push(format!("seq={}: event_hash does not match recomputation — tampered", e.get("seq").map(py_str).unwrap_or_else(|| "None".into())));
        }
        prev = s(e, "event_hash").to_string();
    }
    Ok((problems.is_empty(), problems))
}

fn py_str(v: &Value) -> String {
    match v {
        Value::String(x) => x.clone(),
        Value::Null => "None".into(),
        other => other.to_string(),
    }
}

pub fn verify_record_against_event(fields: &Map<String, Value>, event: &Value, entity_type: &str, entity_id: &str) -> R<bool> {
    let canonical = canonicalize_record(fields, entity_type, entity_id, event.get("format_version").and_then(|v| v.as_i64()).unwrap_or(0))?;
    Ok(payload_hash(&canonical) == s(event, "payload_hash"))
}

/// `json.dump(cp, sort_keys=True)` Python (разделители `, ` и `: `).
fn checkpoint_text(cp: &Map<String, Value>) -> String {
    let mut keys: Vec<&String> = cp.keys().collect();
    keys.sort();
    let mut out = String::from("{");
    for (i, k) in keys.iter().enumerate() {
        if i > 0 {
            out.push_str(", ");
        }
        dumps_str(k, &mut out);
        out.push_str(": ");
        match &cp[*k] {
            Value::String(x) => dumps_str(x, &mut out),
            other => out.push_str(&other.to_string()),
        }
    }
    out.push('}');
    out
}

/// Контрольная точка ВНЕ базы: запись во временный файл, fsync, атомарная замена (обрыв на середине не оставляет половинного файла).
pub fn make_checkpoint(entity_type: &str, events: &[Value], path: &Path, now_secs: i64) -> R<Value> {
    let Some(latest) = events.last() else {
        return Err("ValueError: cannot checkpoint an empty chain".into());
    };
    let mut cp = Map::new();
    cp.insert("entity_type".into(), json!(entity_type));
    cp.insert("seq".into(), latest.get("seq").cloned().unwrap_or(Value::Null));
    cp.insert("event_hash".into(), latest.get("event_hash").cloned().unwrap_or(Value::Null));
    cp.insert("created_at".into(), json!(now_secs));
    let parent = path.parent().filter(|p| !p.as_os_str().is_empty()).unwrap_or(Path::new("."));
    if !parent.exists() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let _ = fs::set_permissions(parent, fs::Permissions::from_mode(0o700));
        }
    }
    let tmp = format!("{}.tmp", path.display());
    {
        let mut f = fs::File::create(&tmp).map_err(|e| e.to_string())?;
        f.write_all(checkpoint_text(&cp).as_bytes()).map_err(|e| e.to_string())?;
        f.flush().map_err(|e| e.to_string())?;
        f.sync_all().map_err(|e| e.to_string())?;
    }
    fs::rename(&tmp, path).map_err(|e| e.to_string())?;
    Ok(Value::Object(cp))
}

pub fn load_checkpoint(path: &Path) -> R<Option<Value>> {
    if !path.exists() {
        return Ok(None);
    }
    let t = fs::read_to_string(path).map_err(|e| e.to_string())?;
    serde_json::from_str(&t).map(Some).map_err(|e| format!("JSONDecodeError: {e}"))
}

/// `(статус, подробность)`: no_checkpoint | ok | ahead | ROLLBACK_SUSPECTED.
pub fn check_for_rollback(current: &[Value], checkpoint_path: &Path) -> R<(String, Option<String>)> {
    let Some(cp) = load_checkpoint(checkpoint_path)? else {
        return Ok(("no_checkpoint".into(), None));
    };
    let Some(latest) = current.last() else {
        return Ok(("ROLLBACK_SUSPECTED".into(), Some("DB has zero events for an entity_type with a prior checkpoint".into())));
    };
    let (ls, cs) = (latest.get("seq").and_then(|v| v.as_i64()).unwrap_or(0), cp.get("seq").and_then(|v| v.as_i64()).unwrap_or(0));
    if ls < cs {
        return Ok(("ROLLBACK_SUSPECTED".into(), Some(format!("DB seq {ls} < checkpoint seq {cs}"))));
    }
    if ls == cs && s(latest, "event_hash") != s(&cp, "event_hash") {
        return Ok(("ROLLBACK_SUSPECTED".into(), Some("same sequence number but different hash — history has diverged".into())));
    }
    if ls == cs {
        return Ok(("ok".into(), None));
    }
    Ok(("ahead".into(), None))
}

/// Вызов по имени для моста тестов: JSON → JSON.
pub fn call(name: &str, args: &Value) -> R<Value> {
    let empty = Map::new();
    let o = args.as_object().unwrap_or(&empty);
    let key = |o: &Map<String, Value>| -> Vec<u8> { s_hex(o.get("key").and_then(|v| v.as_str()).unwrap_or("")) };
    let fields = |o: &Map<String, Value>| -> Map<String, Value> { o.get("fields").and_then(|v| v.as_object()).cloned().unwrap_or_default() };
    let arr = |o: &Map<String, Value>, k: &str| -> Vec<Value> { o.get(k).and_then(|v| v.as_array()).cloned().unwrap_or_default() };
    let text = |o: &Map<String, Value>, k: &str| -> String { o.get(k).map(py_str).unwrap_or_default() };
    let fv = |o: &Map<String, Value>| o.get("format_version").and_then(|v| v.as_i64()).unwrap_or(INTEGRITY_FORMAT_VERSION);
    Ok(match name {
        "canonicalize_value" => canonicalize_value(o.get("value").unwrap_or(&Value::Null))?,
        "canonicalize_record" => json!(String::from_utf8(canonicalize_record(&fields(o), &text(o, "entity_type"), &text(o, "entity_id"), fv(o))?).map_err(|e| e.to_string())?),
        "compute_event_hash" => json!(compute_event_hash(&key(o), fv(o), &text(o, "entity_type"), &text(o, "entity_id"), &text(o, "payload_hash"), &text(o, "previous"))?),
        "append_event" => append_event(&key(o), &arr(o, "chain"), &text(o, "entity_type"), &text(o, "entity_id"), &fields(o), fv(o))?,
        "verify_chain" => {
            let (ok, problems) = verify_chain(&key(o), &arr(o, "events"))?;
            json!([ok, problems])
        }
        "verify_record_against_event" => json!(verify_record_against_event(&fields(o), o.get("event").unwrap_or(&Value::Null), &text(o, "entity_type"), &text(o, "entity_id"))?),
        "make_checkpoint" => make_checkpoint(&text(o, "entity_type"), &arr(o, "events"), Path::new(&text(o, "path")), o.get("now").and_then(|v| v.as_i64()).unwrap_or(0))?,
        "load_checkpoint" => load_checkpoint(Path::new(&text(o, "path")))?.unwrap_or(Value::Null),
        "check_for_rollback" => {
            let (st, d) = check_for_rollback(&arr(o, "current_events"), Path::new(&text(o, "path")))?;
            json!([st, d])
        }
        other => return Err(format!("неизвестная функция {other}")),
    })
}

fn s_hex(h: &str) -> Vec<u8> {
    (0..h.len() / 2).map(|i| u8::from_str_radix(&h[2 * i..2 * i + 2], 16).unwrap_or(0)).collect()
}
